### DISCUSSION TITLE
[Contributor Question] Using Budibase to proxy and authenticate requests to an API

### DISCUSSION THREAD

**Discussion by daawaan4x** [Budibase/budibase#5281] [Help]
Hello, I'm currently trying to extend the integration of the Embedded Map to other GIS Services aside from the XYZ format and I'm also currently trying to add support for securing private GIS Services. However, I ran into technical issues.

My plan is to have the Budibase middleware to proxy private GIS Services to add authentication with its REST API Connector. However, the REST API connectors are fetched with POST requests in the data provider while Leaflet uses GET requests to fetch tiles from GIS Services. I'd like to ask if there's a way to call the REST API connector with a GET request?

I'd appreciate if there is a discord server by Budibase where I can ask for technical details in contributing :>

**Comment by melohagan** [comment]
Hello @daawaan 

I have created a simple address lookup example here: 
[map-export-1649334188392.txt](https://github.com/Budibase/budibase/files/8442761/map-export-1649334188392.txt)

It allows you to enter a street address, and subsequently use an address id to grab the latitude, longitude for the map. 
This free location service was used for demo purposes: https://getaddress.io/

![Screenshot 2022-04-07 at 13 28 15](https://user-images.githubusercontent.com/101575380/162198547-5db9b594-a93e-4b0e-a047-1f1043fb0b9a.png)

![Screenshot 2022-04-07 at 13 28 34](https://user-images.githubusercontent.com/101575380/162198618-aa09341f-4bbe-4227-802f-2239db4fa308.png)

Was this roughly the use case you were trying to achieve? 

You can have a dig through the implementation of the map component here: https://github.com/Budibase/budibase/pull/4856

Let me know if I can help any further, or if there's a particular contribution you'd like to progress.

**Reply by daawaan4x** [reply]
Hello @melohagan, I was referring to finding a way for budibase to act as a middleware to proxy rest APIs and authenticate incoming requests to it. The REST API connectors already work in similar ways but technically it doesn't really proxy APIs since its meant to integrate with the data providers and it uses a different format for the incoming requests from the client. As far as I've been able to understand, requests are "wrapped" in a POST request with a list of the parameter values as the payload, disregarding the original format of the underlying API endpoint and whether it is called with a GET/POST/PUT/DELETE. I tried wandering around the source code for the backend but I honestly do not understand any so I went here for help. In the meantime, I resorted to front-end focused solutions. Specifically, I resorted to using the REST API Connectors and attempting to tinker with the image fetching mechanism of leaflet. I was able to fetch tiles from the proxied GIS Service from `http://appdomain.com/api/v2/queries/<query_id>` but unfortunately, I have yet to tinker with leaflet in order to unwrap and use the image blob in `<img src="{{ createObjectURL(image_blob) }}">` that was wrapped in some json format for data providers. As a sidenote, this feels inefficient so I still think it would be great for Budibase to be able to actually proxy APIs. Not to mention, it is being used as a BaaS with other front-end frameworks like Next.js in some cases.

In my specific case, I've been trying to improve the map component in order to support multiple tile layers. Think of being able to overlay additional raster tiles from private GIS Services. It is simple in theory assuming I won't have to tinker with leaflet anymore. I had a prototype recently, but I scratched it all away because of the security concerns. It can get messy to start working on a UI component right away when the data isn't being fetched securely. I'll try to reconstruct and polish that prototype in a few days and show it here. In the meantime, I want to ask if its possible to add a functionality for budibase atm to act as a middleware to proxy rest APIs and authenticate incoming requests to it?

**Reply by mjashanks** [reply]
Hey @daawaan 

Thanks for the detailed explanation of what you are doing. You are correct when you say that Budibase wraps the REST connectors in POST request, making it difficult to use as a proxy.

Whilst this is a very interesting use case for Budibase, it's the first time it's been requested. So, we probably won't be adding this anytime soon unless there is considerable demand from others as well. 

Thank you so much for offering to contribute this, but please don't spend time on it as it's outside of scope of the product, as far as I can see at the moment. Happy to be proven wrong, if others want to weigh in and give some convincing uses for this :)

We'd love you to be involved in other contributions that are already in our roadmap & issues list.

**Reply by daawaan4x** [reply]
Hello @mjashanks, thank you very much! I'll be willing to look in other projects in the roadmap. In the meantime, I'll finish improvising the `EmbeddedMap` to allow multiple tile layers (with the front-end focused solution I mentioned) since our organization needs it anyway
